### DISCUSSION TITLE
Provide distributed-systems migration functionality / documentation

### DISCUSSION THREAD

**Discussion by DarkTrick** [sqlalchemy/sqlalchemy#10261] [Ideas]
## Background
I feel sqlalchemy expects the whole system to upgrade at one single point in type. However, for distributed systems this is not possible. For these systems, having a mechanism to handle the situation of different model definitions at the same time would be very helpful, I think.

## Features I have in mind:

1. Column aliasing
2. Provide column access logic
3. Provide model versioning

Explanation:
1. I guess no explanation needed.
2. Basically like a *setter* function for model attributes. 
   - Use case 1: Column moved to a different table. The setter function could forward the *set* action to the other table while parts of the distributed system are not updated, yet.
   - Use case 2: Column type changed. The setter function could transform data of the old type to the new type, while parts of the distributed system are not updated, yet.
3. Provide a way to manage multiple versions of a table definition. The idea is to have one model specifying the actual current structure of the database, while there can be parts of tables, that specify an older version of a table. (This goes hand in hand with alembic, I guess. Perhaps here is not even the right place to state this).

## Implementation

While aliasing would be nice sugar, I think these feature do not necessarily need to be *implemented in code* by sqlalchemy. It might be enough to add a chapter in the documentation addressing this issue and showing a best-practice pattern how to solve it.

**Comment by zzzeek** [comment]
is this question more concerning schema migrations , i.e. alembic?   I dont see what features in SQLAlchemy itself are needed for the general concept of  "distributed systems".    

for an example of a large distributed system built on SQLAlchemy and Alembic see [openstack](https://www.openstack.org/).  I work with this system as my day job so I can answer any questions you may have regarding its architecture and how it approaches various issues.

**Reply by zzzeek** [reply]
Specifically:

> 1. column aliasing

for the ORM this is achieved using [synonym](https://docs.sqlalchemy.org/en/20/orm/mapped_attributes.html#synonyms), [column_property](https://docs.sqlalchemy.org/en/20/orm/mapped_sql_expr.html#using-column-property) and [hybrid](https://docs.sqlalchemy.org/en/20/orm/mapped_sql_expr.html#using-a-hybrid)

> Basically like a setter function for model attributes.

[hybrid](https://docs.sqlalchemy.org/en/20/orm/mapped_sql_expr.html#using-a-hybrid)

>    Use case 1: Column moved to a different table. The setter function could forward the set action to the other table while parts of the distributed system are not updated, yet.

[hybrid](https://docs.sqlalchemy.org/en/20/orm/mapped_sql_expr.html#using-a-hybrid)


>   Use case 2: Column type changed. The setter function could transform data of the old type to the new type, while parts of the distributed system are not updated, yet.

mostly [hybrid](https://docs.sqlalchemy.org/en/20/orm/mapped_sql_expr.html#using-a-hybrid) again here, or if you want to get at it at the type level, [TypeDecorator](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator)

> Provide a way to manage multiple versions of a table definition. The idea is to have one model specifying the actual current structure of the database, while there can be parts of tables, that specify an older version of a table.

this is usually just the columns from the "previous" version that are still on the `Table` definition.  you will usually want these on your model in any case since you would need to sync data to and from those columns, unless you are using triggers on the server to do this, in which case you dont need to map those columns at all.

**Comment by DarkTrick** [comment]
zzzeek, thank you very much for your quick reply! I really appreciate it! 
I'll take a look at everything you mentioned and will close this discussion for now.
